### PR TITLE
solution #3: find the largest in a list

### PR DESCRIPTION
**Pull request by Khadijaramlawi**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [ ] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in any docstrings or tests
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_find_largest.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+Author : Khadija  Ramlawi
+"""
+
+import unittest
+from solutions.find_largest import find_largest
+
+
+class TestFindLargest(unittest.TestCase):
+    def test_positive_numbers(self):
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:14]: Add class docstring.

File: solutions/tests/test_find_largest.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        # List with positive numbers
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:15]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertEqual(find_largest([1, 2, 3]), 3)
+
+    def test_mixed_numbers(self):
+        # List with both positive and negative numbers
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:19]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertEqual(find_largest([-10, 0, 5, 20]), 20)
+
+    def test_all_negative_numbers(self):
+        # List with all negative numbers
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:23]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertEqual(find_largest([-5, -1, -10]), -1)
+
+    def test_single_element(self):
+        # List with a single element
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:27]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertEqual(find_largest([42]), 42)
+
+    def test_large_numbers(self):
+        # List with large numbers
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:31]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertEqual(find_largest([1000000, 5000000, 2000000]), 5000000)
+
+    def test_empty_list(self):
+        # Empty list should raise an AssertionError
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:35]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        with self.assertRaises(AssertionError):
+            find_largest([])
+
+    def test_non_numeric_values(self):
+        # List with non-numeric values should raise an AssertionError
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:40]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        with self.assertRaises(AssertionError):
+            find_largest([1, 2, "three"])
+
+    def test_invalid_input_non_list(self):
+        # Non-list input should raise an AssertionError
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:45]: Remove comment and include test function docstring.

File: solutions/tests/test_find_largest.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        with self.assertRaises(AssertionError):
+            find_largest(123)
+
+    def test_invalid_input_none(self):
+        # None as input should raise an AssertionError
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:50]: Remove comment and include test function docstring.

File: solutions/find_largest.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
+"""
```
Comment (suhrobmuboraksho) [inline solutions/find_largest.py:1]: Not sure if needed or not.
You could include
#!/usr/bin/env python3
#-*- coding: utf-8 -*-

at the beginning of the file.

File: solutions/tests/test_find_largest.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
+"""
```
Comment (suhrobmuboraksho) [inline solutions/tests/test_find_largest.py:1]: Not sure if needed or not.
You could include
#!/usr/bin/env python3
#-- coding: utf-8 --

at the beginning of the file.

**Review by suhrobmuboraksho — COMMENTED** [review]
Left my comments.
Module docsting present
Doctest runs without mistake.
All 9 unittests are passed.
Edge cases are taken into account.
Missing function level docstrings in test file.
Comments are in the code.

**Comment by Khadijaramlawi** [comment]
> Left my comments. Module docstring present Doctest runs without mistake. All 9 unittests are passed. Edge cases are taken into account. Missing function level docstrings in test file. Comments are in the code.

@suhrobmuboraksho  Thank you so much Suhrob for valuable comments! I did post of it and I didn't pay attention previously  to the doctorings in the unites. I thought Evan used comments instead. 
I hope all is fixed now. Thank you!

**Review by suhrobmuboraksho — APPROVED** [review]
Module, class and function level docstrings are present.
Unittests include edge cases.
Unittests pass all checks.
Doctest didn't reveal any issues.
Code is well written and structured.

**Review by Papsik09 — APPROVED** [review]
Well done. Resolved all comments. Keep it up!!!
